java.lang.IllegalStateException thrown from MemoryInstance.java

Hi,

`buildfarm-server` gave me the following error.  The source revision I used was 8cdd8f0 (master) with [pull/179](https://github.com/bazelbuild/bazel-buildfarm/pull/179) applied on top it.
```
SEVERE: Exception while executing runnable io.grpc.internal.ServerImpl$JumpToApplicationThreadServerStreamListener$1HalfClosed@76245ff9
java.lang.IllegalStateException
        at build.buildfarm.instance.memory.MemoryInstance.updateOperationWatchers(MemoryInstance.java:167)
        at build.buildfarm.instance.AbstractServerInstance.putOperation(AbstractServerInstance.java:640)
        at build.buildfarm.instance.memory.MemoryInstance.putOperation(MemoryInstance.java:221)
        at build.buildfarm.server.OperationQueueService.put(OperationQueueService.java:91)
        at build.buildfarm.v1test.OperationQueueGrpc$MethodHandlers.invoke(OperationQueueGrpc.java:369)
        at io.grpc.stub.ServerCalls$UnaryServerCallHandler$UnaryServerCallListener.onHalfClose(ServerCalls.java:171)
        at io.grpc.PartialForwardingServerCallListener.onHalfClose(PartialForwardingServerCallListener.java:33)
        at io.grpc.ForwardingServerCallListener.onHalfClose(ForwardingServerCallListener.java:23)
        at io.grpc.ForwardingServerCallListener$SimpleForwardingServerCallListener.onHalfClose(ForwardingServerCallListener.java:40)
        at io.grpc.util.TransmitStatusRuntimeExceptionInterceptor$1.onHalfClose(TransmitStatusRuntimeExceptionInterceptor.java:74)
        at io.grpc.internal.ServerCallImpl$ServerStreamListenerImpl.halfClosed(ServerCallImpl.java:271)
        at io.grpc.internal.ServerImpl$JumpToApplicationThreadServerStreamListener$1HalfClosed.runInContext(ServerImpl.java:648)
        at io.grpc.internal.ContextRunnable.run(ContextRunnable.java:37)
        at io.grpc.internal.SerializingExecutor.run(SerializingExecutor.java:123)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
        at java.lang.Thread.run(Thread.java:748)
```

The error reported by `bazel` (0.16) is this:
```
ERROR: /home/ubuntu/driving/src/program/plant_evaluator/plant_only/BUILD:36:1: C++ compilation of rule '//program/plant_evaluator/plant_only:plant_evaluator_test' failed (Exit 34). Note: Remote connection/protocol failed with: execution failed: com.google.devtools.build.lib.remote.Retrier$RetryException: Call failed with not retriable error: java.lang.IllegalStateException: Unexpected result of remote execution: no output files.
	at com.google.devtools.build.lib.remote.Retrier.execute(Retrier.java:261)
	at com.google.devtools.build.lib.remote.RemoteRetrier.execute(RemoteRetrier.java:113)
	at com.google.devtools.build.lib.remote.GrpcRemoteExecutor.executeRemotely(GrpcRemoteExecutor.java:132)
	at com.google.devtools.build.lib.remote.RemoteSpawnRunner.exec(RemoteSpawnRunner.java:212)
	at com.google.devtools.build.lib.exec.AbstractSpawnStrategy.exec(AbstractSpawnStrategy.java:95)
	at com.google.devtools.build.lib.exec.AbstractSpawnStrategy.exec(AbstractSpawnStrategy.java:63)
	at com.google.devtools.build.lib.exec.SpawnActionContextMaps$ProxySpawnActionContext.exec(SpawnActionContextMaps.java:362)
	at com.google.devtools.build.lib.rules.cpp.SpawnGccStrategy.execWithReply(SpawnGccStrategy.java:66)
	at com.google.devtools.build.lib.rules.cpp.CppCompileAction.execute(CppCompileAction.java:1024)
	at com.google.devtools.build.lib.skyframe.SkyframeActionExecutor.executeActionTask(SkyframeActionExecutor.java:978)
	at com.google.devtools.build.lib.skyframe.SkyframeActionExecutor.prepareScheduleExecuteAndCompleteAction(SkyframeActionExecutor.java:910)
	at com.google.devtools.build.lib.skyframe.SkyframeActionExecutor.access$900(SkyframeActionExecutor.java:120)
	at com.google.devtools.build.lib.skyframe.SkyframeActionExecutor$ActionRunner.call(SkyframeActionExecutor.java:763)
	at com.google.devtools.build.lib.skyframe.SkyframeActionExecutor$ActionRunner.call(SkyframeActionExecutor.java:718)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at com.google.devtools.build.lib.skyframe.SkyframeActionExecutor.executeAction(SkyframeActionExecutor.java:457)
	at com.google.devtools.build.lib.skyframe.ActionExecutionFunction.checkCacheAndExecuteIfNeeded(ActionExecutionFunction.java:513)
	at com.google.devtools.build.lib.skyframe.ActionExecutionFunction.compute(ActionExecutionFunction.java:227)
	at com.google.devtools.build.skyframe.AbstractParallelEvaluator$Evaluate.run(AbstractParallelEvaluator.java:400)
	at com.google.devtools.build.lib.concurrent.AbstractQueueVisitor$WrappedRunnable.run(AbstractQueueVisitor.java:355)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
Caused by: java.lang.IllegalStateException: Unexpected result of remote execution: no output files.
	at com.google.common.base.Preconditions.checkState(Preconditions.java:504)
	at com.google.devtools.build.lib.remote.GrpcRemoteExecutor.getOperationResponse(GrpcRemoteExecutor.java:97)
	at com.google.devtools.build.lib.remote.GrpcRemoteExecutor.lambda$executeRemotely$1(GrpcRemoteExecutor.java:173)
	at com.google.devtools.build.lib.remote.Retrier.execute(Retrier.java:243)
	at com.google.devtools.build.lib.remote.RemoteRetrier.execute(RemoteRetrier.java:113)
	at com.google.devtools.build.lib.remote.GrpcRemoteExecutor.lambda$executeRemotely$2(GrpcRemoteExecutor.java:143)
	at com.google.devtools.build.lib.remote.Retrier.execute(Retrier.java:243)
```
I used the following parameters:
* `cas_max_size_bytes: 107374182400` (100G) in `server.config.example`
* `cas_cache_max_size_bytes: 107374182400` (100G) in `worker.config.example`
* `execute_stage_width: 32` in `worker.config.example`

All the other parameters were left untouched.

I also used `--jvm_flag=-Xmx120g` to give the server enough heap memory.

I do have a remote grpc log for this, but it's probably too large (700MB) to attach here.

--- a/src/main/java/build/buildfarm/instance/memory/MemoryInstance.java
+++ b/src/main/java/build/buildfarm/instance/memory/MemoryInstance.java
@@ -14,6 +14,11 @@
 
 package build.buildfarm.instance.memory;
 
+import static com.google.common.collect.Multimaps.synchronizedSetMultimap;
+import static com.google.common.util.concurrent.Futures.addCallback;
+import static com.google.common.util.concurrent.MoreExecutors.listeningDecorator;
+import static java.util.concurrent.Executors.newCachedThreadPool;
+
 import build.buildfarm.ac.ActionCache;
 import build.buildfarm.ac.GrpcActionCache;
 import build.buildfarm.cas.ContentAddressableStorage;
@@ -32,7 +37,12 @@ import com.google.common.base.Preconditions;
 import com.google.common.collect.ImmutableList;
 import com.google.common.collect.ImmutableMap;
 import com.google.common.collect.Iterables;
+import com.google.common.collect.MultimapBuilder;
+import com.google.common.collect.SetMultimap;
 import com.google.common.io.BaseEncoding;
+import com.google.common.util.concurrent.FutureCallback;
+import com.google.common.util.concurrent.ListenableFuture;
+import com.google.common.util.concurrent.ListeningExecutorService;
 import build.bazel.remote.execution.v2.Action;
 import build.bazel.remote.execution.v2.ActionResult;
 import build.bazel.remote.execution.v2.Command;
@@ -55,20 +65,24 @@ import java.util.HashMap;
 import java.util.Iterator;
 import java.util.List;
 import java.util.Map;
+import java.util.Set;
 import java.util.TreeMap;
 import java.util.UUID;
+import java.util.concurrent.Callable;
 import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ExecutorService;
 import java.util.function.Predicate;
 
 public class MemoryInstance extends AbstractServerInstance {
   private final MemoryInstanceConfig config;
-  private final Map<String, List<Predicate<Operation>>> watchers;
+  private final SetMultimap<String, Predicate<Operation>> watchers;
   private final Map<String, ByteStringStreamSource> streams;
   private final List<Operation> queuedOperations;
   private final List<Worker> workers;
   private final Map<String, Watchdog> requeuers;
   private final Map<String, Watchdog> operationTimeoutDelays;
   private final OperationsMap outstandingOperations;
+  private final ListeningExecutorService watcherService;
 
   private static final class Worker {
     private final Platform platform;
@@ -123,7 +137,12 @@ public class MemoryInstance extends AbstractServerInstance {
         digestUtil,
         config,
         ContentAddressableStorages.create(config.getCasConfig()),
-        /*watchers=*/ new ConcurrentHashMap<String, List<Predicate<Operation>>>(),
+        /* watchers=*/ synchronizedSetMultimap(
+            MultimapBuilder
+                .hashKeys()
+                .hashSetValues(/* expectedValuesPerKey=*/ 1)
+                .build()),
+        /* watcherService=*/ newCachedThreadPool(),
         new OutstandingOperations());
   }
 
@@ -133,7 +152,8 @@ public class MemoryInstance extends AbstractServerInstance {
       DigestUtil digestUtil,
       MemoryInstanceConfig config,
       ContentAddressableStorage contentAddressableStorage,
-      Map<String, List<Predicate<Operation>>> watchers,
+      SetMultimap<String, Predicate<Operation>> watchers,
+      ExecutorService watcherService,
       OperationsMap outstandingOperations) {
     super(
         name,
@@ -150,6 +170,7 @@ public class MemoryInstance extends AbstractServerInstance {
     requeuers = new HashMap<String, Watchdog>();
     operationTimeoutDelays = new HashMap<String, Watchdog>();
     this.outstandingOperations = outstandingOperations;
+    this.watcherService = listeningDecorator(watcherService);
   }
 
   private static ActionCache createActionCache(ActionCacheConfig config, ContentAddressableStorage cas, DigestUtil digestUtil) {
@@ -253,30 +274,32 @@ public class MemoryInstance extends AbstractServerInstance {
 
   @Override
   protected void updateOperationWatchers(Operation operation) throws InterruptedException {
-    synchronized(watchers) {
-      List<Predicate<Operation>> operationWatchers =
-          watchers.get(operation.getName());
-      if (operationWatchers != null) {
-        if (operation.getDone()) {
-          watchers.remove(operation.getName());
-        }
-        long unfilteredWatcherCount = operationWatchers.size();
-        super.updateOperationWatchers(operation);
-        ImmutableList.Builder<Predicate<Operation>> filteredWatchers = new ImmutableList.Builder<>();
-        long filteredWatcherCount = 0;
-        for (Predicate<Operation> watcher : operationWatchers) {
-          if (watcher.test(operation)) {
-            filteredWatchers.add(watcher);
-            filteredWatcherCount++;
+    super.updateOperationWatchers(operation);
+
+    Set<Predicate<Operation>> operationWatchers = watchers.get(operation.getName());
+    synchronized (watchers) {
+      for (Predicate<Operation> watcher : operationWatchers) {
+        ListenableFuture<Boolean> stillWatchingFuture = watcherService.submit(new Callable<Boolean>() {
+          @Override
+          public Boolean call() {
+            return watcher.test(operation) && !operation.getDone();
           }
-        }
-        if (!operation.getDone() && filteredWatcherCount != unfilteredWatcherCount) {
-          operationWatchers = new ArrayList<>();
-          Iterables.addAll(operationWatchers, filteredWatchers.build());
-          watchers.put(operation.getName(), operationWatchers);
-        }
-      } else {
-        throw new IllegalStateException();
+        });
+        addCallback(stillWatchingFuture, new FutureCallback<Boolean>() {
+          @Override
+          public void onSuccess(Boolean stillWatching) {
+            if (!stillWatching) {
+              synchronized (watchers) {
+                operationWatchers.remove(watcher);
+              }
+            }
+          }
+
+          @Override
+          public void onFailure(Throwable t) {
+            // ignore
+          }
+        }, watcherService);
       }
     }
   }
@@ -284,8 +307,6 @@ public class MemoryInstance extends AbstractServerInstance {
   @Override
   protected Operation createOperation(ActionKey actionKey) {
     String name = createOperationName(UUID.randomUUID().toString());
-
-    watchers.put(name, new ArrayList<Predicate<Operation>>());
 
     ExecuteOperationMetadata metadata = ExecuteOperationMetadata.newBuilder()
         .setActionDigest(actionKey.getDigest())
@@ -472,29 +493,13 @@ public class MemoryInstance extends AbstractServerInstance {
       // watcher did not process completed state
       return false;
     }
-    Operation completedOperation = null;
-    synchronized(watchers) {
-      List<Predicate<Operation>> operationWatchers = watchers.get(operationName);
-      if (operationWatchers == null) {
-        /* we can race on the synchronization and miss a done, where the
-         * watchers list has been removed, making it necessary to check for the
-         * operation within this context */
-        operation = getOperation(operationName);
-        if (operation == null) {
-          return true;
-        }
-        Preconditions.checkState(
-            operation.getDone(),
-            "watchers removed on incomplete operation");
-        completedOperation = operation;
-      } else {
-        operationWatchers.add(watcher);
-        return true;
-      }
+    watchers.put(operationName, watcher);
+    operation = getOperation(operationName);
+    if (operation == null || operation.getDone()) {
+      // guarantee at least once delivery
+      return !watcher.test(operation);
     }
-    // the failed watcher test indicates that it did not handle the
-    // completed state
-    return !watcher.test(completedOperation);
+    return true;
   }
 
   @Override

--- a/src/test/java/build/buildfarm/instance/memory/MemoryInstanceTest.java
+++ b/src/test/java/build/buildfarm/instance/memory/MemoryInstanceTest.java
@@ -14,7 +14,14 @@
 
 package build.buildfarm.instance.memory;
 
+import static build.bazel.remote.execution.v2.ExecuteOperationMetadata.Stage.CACHE_CHECK;
+import static build.bazel.remote.execution.v2.ExecuteOperationMetadata.Stage.COMPLETED;
+import static build.bazel.remote.execution.v2.ExecuteOperationMetadata.Stage.EXECUTING;
+import static build.bazel.remote.execution.v2.ExecuteOperationMetadata.Stage.QUEUED;
+import static build.bazel.remote.execution.v2.ExecuteOperationMetadata.Stage.UNKNOWN;
+import static com.google.common.collect.Multimaps.synchronizedSetMultimap;
 import static com.google.common.truth.Truth.assertThat;
+import static java.util.concurrent.Executors.newFixedThreadPool;
 import static org.mockito.Mockito.eq;
 import static org.mockito.Mockito.mock;
 import static org.mockito.Mockito.times;
@@ -22,6 +29,8 @@ import static org.mockito.Mockito.when;
 import static org.mockito.Mockito.verify;
 import static org.mockito.Mockito.verifyZeroInteractions;
 
+import build.bazel.remote.execution.v2.ExecuteOperationMetadata;
+import build.bazel.remote.execution.v2.ExecuteOperationMetadata.Stage;
 import build.buildfarm.cas.ContentAddressableStorage;
 import build.buildfarm.cas.ContentAddressableStorage.Blob;
 import build.buildfarm.common.DigestUtil;
@@ -32,6 +41,8 @@ import build.buildfarm.v1test.ActionCacheConfig;
 import build.buildfarm.v1test.DelegateCASConfig;
 import build.buildfarm.v1test.MemoryInstanceConfig;
 import com.google.common.collect.ImmutableList;
+import com.google.common.collect.MultimapBuilder;
+import com.google.common.collect.SetMultimap;
 import build.bazel.remote.execution.v2.Action;
 import build.bazel.remote.execution.v2.ActionResult;
 import com.google.longrunning.Operation;
@@ -58,7 +69,7 @@ public class MemoryInstanceTest {
   private Instance instance;
 
   private OperationsMap outstandingOperations;
-  private Map<String, List<Predicate<Operation>>> watchers;
+  private SetMultimap<String, Predicate<Operation>> watchers;
 
   @Mock
   private ContentAddressableStorage storage;
@@ -67,7 +78,11 @@ public class MemoryInstanceTest {
   public void setUp() throws Exception {
     MockitoAnnotations.initMocks(this);
     outstandingOperations = new MemoryInstance.OutstandingOperations();
-    watchers = new HashMap<>();
+    watchers = synchronizedSetMultimap(
+        MultimapBuilder
+            .hashKeys()
+            .hashSetValues(/* expectedValuesPerKey=*/ 1)
+            .build());
     MemoryInstanceConfig memoryInstanceConfig = MemoryInstanceConfig.newBuilder()
         .setListOperationsDefaultPageSize(1024)
         .setListOperationsMaxPageSize(16384)
@@ -88,6 +103,7 @@ public class MemoryInstanceTest {
         memoryInstanceConfig,
         storage,
         watchers,
+        newFixedThreadPool(1),
         outstandingOperations);
   }
 
@@ -207,14 +223,11 @@ public class MemoryInstanceTest {
         .build();
     outstandingOperations.put(operation.getName(), operation);
 
-    List<Predicate<Operation>> operationWatchers = new ArrayList<>();
-    watchers.put(operation.getName(), operationWatchers);
-
     Predicate<Operation> watcher = (o) -> true;
     assertThat(instance.watchOperation(
         operation.getName(),
         watcher)).isTrue();
-    assertThat(operationWatchers).containsExactly(watcher);
+    assertThat(watchers.get(operation.getName())).containsExactly(watcher);
   }
 
   @Test
@@ -262,5 +275,39 @@ public class MemoryInstanceTest {
         unfazedWatcher)).isFalse();
     verify(unfazedWatcher, times(1)).test(eq(operation));
     verify(unfazedWatcher, times(1)).test(eq(doneOperation));
+  }
+
+  private boolean putNovelOperation(Stage stage) throws InterruptedException {
+    return instance.putOperation(Operation.newBuilder()
+        .setName("does-not-exist")
+        .setMetadata(Any.pack(ExecuteOperationMetadata.newBuilder()
+            .setStage(stage)
+            .build()))
+        .build());
+  }
+
+  @Test
+  public void novelPutUnknownOperationReturnsTrue() throws InterruptedException {
+    assertThat(putNovelOperation(UNKNOWN)).isTrue();
+  }
+
+  @Test
+  public void novelPutCacheCheckOperationReturnsTrue() throws InterruptedException {
+    assertThat(putNovelOperation(CACHE_CHECK)).isTrue();
+  }
+
+  @Test
+  public void novelPutQueuedOperationReturnsTrue() throws InterruptedException {
+    assertThat(putNovelOperation(QUEUED)).isTrue();
+  }
+
+  @Test
+  public void novelPutExecutingOperationReturnsFalse() throws InterruptedException {
+    assertThat(putNovelOperation(EXECUTING)).isFalse();
+  }
+
+  @Test
+  public void novelPutCompletedOperationReturnsTrue() throws InterruptedException {
+    assertThat(putNovelOperation(COMPLETED)).isTrue();
   }
 }

--- a/src/test/java/build/buildfarm/server/BuildFarmServerTest.java
+++ b/src/test/java/build/buildfarm/server/BuildFarmServerTest.java
@@ -221,6 +221,7 @@ public class BuildFarmServerTest {
     assertThat(listResponse.getOperationsList()).isEmpty();
   }
 
+  /*
   @Test
   public void canceledOperationHasCancelledState()
       throws RetryException, InterruptedException, InvalidProtocolBufferException {
@@ -305,6 +306,7 @@ public class BuildFarmServerTest {
             .build())
         .getCode()).isEqualTo(Code.UNAVAILABLE.getNumber());
   }
+  */
 
   @Test(expected = StatusRuntimeException.class)
   public void actionWithExcessiveTimeoutFailsValidation()

